src/fluree/db/query/fql/parse.cljc

```diff
@@ -0,0 +1,428 @@
+   (parse-sid x))
+
+  ([x context]
+   (if-let [parsed (parse-subject x)]
+     parsed
+     (when context
+       (parse-iri x context)))))
+
+(defn parse-subject-pattern
+  [s-pat context]
+  (when s-pat
+    (or (parse-variable s-pat)
+        (parse-pred-ident s-pat)
+        (parse-subject s-pat context)
+        (throw (ex-info (str "Subject values in where statement must be integer subject IDs or two-tuple identies. "
+                             "Provided: " s-pat ".")
+                        {:status 400 :error :db/invalid-query})))))
+
+(defn parse-class-predicate
+  [x]
+  (when (rdf-type? x)
+    (where/->value const/$rdf:type)))
+
+(defn parse-iri-predicate
+  [x]
+  (when (= "@id" x)
+    (where/->predicate const/$iri)))
+
+(defn iri->pred-id
+  [iri db context]
+  (let [full-iri (json-ld/expand-iri iri context)]
+    (dbproto/-p-prop db :id full-iri)))
+
+(defn iri->pred-id-strict
+  [iri db context]
+  (or (iri->pred-id iri db context)
+      (throw (ex-info (str "Invalid predicate: " iri)
+                      {:status 400 :error :db/invalid-query}))))
+
+(defn parse-recursion-predicate
+  [x db context]
+  (when-let [[p-iri recur-n] (recursion-predicate x context)]
+    (let [iri     (iri->pred-id-strict p-iri db context)
+          recur-n (or recur-n util/max-integer)]
+      (where/->predicate iri recur-n))))
+
+(defn parse-full-text-predicate
+  [x db context]
+  (when (and (string? x)
+             (str/starts-with? x "fullText:"))
+    (-> x
+        (subs 9)
+        (iri->pred-id-strict db context)
+        where/->full-text)))
+
+(defn parse-predicate-id
+  [x db context]
+  (-> x
+      (iri->pred-id-strict db context)
+      where/->value))
+
+(defn parse-predicate-pattern
+  [p-pat db context]
+  (or (parse-iri-predicate p-pat)
+      (parse-class-predicate p-pat)
+      (parse-variable p-pat)
+      (parse-recursion-predicate p-pat db context)
+      (parse-full-text-predicate p-pat db context)
+      (parse-predicate-id p-pat db context)))
+
+(defn parse-class
+  [o-iri db context]
+  (if-let [id (iri->pred-id o-iri db context)]
+    (parse-sid id)
+    (throw (ex-info (str "Undefined RDF type specified: " (json-ld/expand-iri o-iri context))
+                    {:status 400 :error :db/invalid-query}))))
+
+(defn parse-object-pattern
+  [o-pat context]
+  (or (parse-variable o-pat)
+      (parse-pred-ident o-pat)
+      (where/->value o-pat)))
+
+(defmulti parse-pattern
+  (fn [pattern vars db context]
+    (if (map? pattern)
+      (->> pattern keys first)
+      :triple)))
+
+(defn filter-pattern?
+  [x]
+  (and (map? x)
+       (-> x keys first (= :filter))))
+
+(defn parse-filter-maps
+  [vars filters]
+  (let [vars (set vars)]
+    (->> filters
+         (mapcat vals)
+         flatten
+         (map (fn [f-str]
+                (parse-filter-function f-str vars)))
+         (reduce (fn [m fltr]
+                   (let [var-name (::where/var fltr)]
+                     (update m var-name (fn [var-fltrs]
```

Comment:
This is elegant - I like how we just build up a sequence of filter functions per var.